controllers/drplacementcontrol.go

```diff
@@ -0,0 +1,1372 @@
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/go-logr/logr"
+	ocmworkv1 "github.com/open-cluster-management/api/work/v1"
+	plrv1 "github.com/open-cluster-management/multicloud-operators-placementrule/pkg/apis/apps/v1"
+	errorswrapper "github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	rmn "github.com/ramendr/ramen/api/v1alpha1"
+	rmnutil "github.com/ramendr/ramen/controllers/util"
+)
+
+type DRPCInstance struct {
+	reconciler           *DRPlacementControlReconciler
+	ctx                  context.Context
+	log                  logr.Logger
+	instance             *rmn.DRPlacementControl
+	drPolicy             *rmn.DRPolicy
+	needStatusUpdate     bool
+	mcvRequestInProgress bool
+	userPlacementRule    *plrv1.PlacementRule
+	drpcPlacementRule    *plrv1.PlacementRule
+	vrgs                 map[string]*rmn.VolumeReplicationGroup
+	mwu                  rmnutil.MWUtil
+}
+
+func (d *DRPCInstance) startProcessing() bool {
+	d.log.Info("Starting to process placement")
+
+	requeue := true
+
+	done, err := d.processPlacement()
+	if err != nil {
+		d.log.Info("Process placement", "error", err.Error())
+
+		return requeue
+	}
+
+	if d.needStatusUpdate || d.statusUpdateTimeElapsed() {
+		if err := d.updateDRPCStatus(); err != nil {
+			d.log.Error(err, "failed to update status")
+
+			return requeue
+		}
+	}
+
+	requeue = !done
+	d.log.Info("Completed processing placement", "requeue", requeue)
+
+	return requeue
+}
+
+func (d *DRPCInstance) processPlacement() (bool, error) {
+	d.log.Info("Process DRPC Placement", "DRAction", d.instance.Spec.Action)
+
+	switch d.instance.Spec.Action {
+	case rmn.ActionFailover:
+		return d.runFailover()
+	case rmn.ActionRelocate:
+		return d.runRelocate()
+	}
+
+	// Not a failover or a relocation.  Must be an initial deployment.
+	return d.runInitialDeployment()
+}
+
+func (d *DRPCInstance) runInitialDeployment() (bool, error) {
+	d.log.Info("Running initial deployment")
+
+	const done = true
+
+	homeCluster, homeClusterNamespace := d.getHomeCluster()
+
+	if homeCluster == "" {
+		err := fmt.Errorf("PreferredCluster not set and unable to find home cluster in DRPCPlacementRule (%v)",
+			d.drpcPlacementRule)
+		// needStatusUpdate is not set. Still better to capture the event to report later
+		rmnutil.ReportIfNotPresent(d.reconciler.eventRecorder, d.instance, corev1.EventTypeWarning,
+			rmnutil.EventReasonDeployFail, err.Error())
+
+		return !done, err
+	}
+
+	d.log.Info(fmt.Sprintf("Using homeCluster %s for initial deployment, uPlRule Decision %v",
+		homeCluster, d.userPlacementRule.Status.Decisions))
+
+	// Check if we already deployed in the homeCluster or elsewhere
+	clusterName, deployed := d.isDeployed(homeCluster)
+	if deployed && clusterName != homeCluster {
+		return done, nil
```

Comment:
If cluster name is not the homeCluster, should we finish the reconciliation to the targetCluster, i.e ensure user placement is updated etc.?

Reply:
I think if we do that, we would be (silently) violating the user configuration. It is actually better to just not do anything and let the user intervene.  We should definitely log after line 114

Reply:
Understood the intent. Deploy will act to initially deploy, if spec is changed by user in arbitrary ways, deployment will stall, and has to be recovered with an action.